README
Dist::Zilla::Plugin::MetaProvides

This Distribution Contains a small bundle of plugins for various ways of
populating the "META.yml" that is built with your distribution.

The initial reason for this is due to stuff that uses MooseX::Declare
style class definitions not being parseable by many tools upstream, so
this is here to cover this problem by defining it in the metadata.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Dist-Zilla-Plugin-MetaProvides

2.002005 2017-03-01T12:19:21Z cc8c243

2.002004 2017-03-01T11:57:55Z fd9040a
 [Bugfix]
 - Update tooling to avoid .-in-@INC issues
 - Closes https://github.com/kentnl/Dist-Zilla-Plugin-MetaProvides/issues/7

 [Dependencies::Stats]
 - Dependencies changed since 2.002003, see misc/*.deps* for details
 - develop: +1 ↑1

2.002003 2016-07-29T22:25:40Z c9860d3
 - STABLE: This release includes all changes previously stated between 2.002000 and 2.002000 as stable.
 - SHORTLIST: Better logging about possible issues, Fewer dependencies, Enhanced metaconfig data, MetaNoIndex
   application now correct.

 [Dependencies::Stats]
 - Dependencies changed since 2.002002, see misc/*.deps* for details
 - develop: (suggests: ↑1)

2.002002 2016-07-07T11:55:30Z 2a9e822
 - Adds logging warnings for Provides entries that don't map to shipped files. This should not be a problem for anyone
   unless you're using custom MetaProvides.
 - Warn about cuckooed packages at the "provides-to-dzil" level. This enables automatic exposure of all packages where
   "$filename" !~ "/Module/Name.pm" as being potentially problematic.

 [Dependencies::Stats]
 - Dependencies changed since 2.002001, see misc/*.deps* for details
 - test: +1

2.002001 2016-06-29T06:09:34Z 2be5312
 - Remove usage of ConfigDumper and inline dump_config code instead.
 - use '$' on front of role-injected '$VERSION' identifier to be clearer.
 - Remove usage of my dztest testing util.

 [Dependencies::Stats]
 - Dependencies changed since 2.002000, see misc/*.deps* for details
 - runtime: -1
 - test: +1 -1 (recommends: +1)

2.002000 2016-04-22T14:20:31Z 8cd4f33
 - IMPORTANT: Previously, MetaNoIndex compatibility implemented the "namespace" feature in contradiction of
   https://metacpan.org/pod/CPAN::Meta::Spec#namespace , which means packages that exactly matched the "namespace"
   token were excluded, when they should have not been excluded. This release resolves this issue, as reported at
   https://github.com/rjbs/Dist-Zilla/issues/204

2.001011 2016-01-24T11:24:45Z b20733b
 [Dependencies::Stats]
 - Dependencies changed since 2.001010, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +1

 [Documentation]
 - Enhanced Synopsis documentation as per #4

 [Features]
 - Customized Croak message for people who accidentally use [MetaProvides] directly. Closes #4

2.001010 2015-10-09T07:02:08Z d62ab63
 [Dependencies::Stats]
 - Dependencies changed since 2.001002, see misc/*.deps* for details
 - configure: (recommends: ↓1)
 - develop: +8 ↑3 -1 (recommends: ↓1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: (recommends: ↓2)

 [Documentation]
 - Remove confusing quick reference and try to find a better way another day
 - kentnl/Dist-Zilla-Plugin-MetaProvides-Package#4

 [Features]
 - Experimentally injecting the version of the role into `dump_config` data.
 - Part of request #3

2.001002 2014-12-10T10:16:51Z 6baccf2
 [Dependencies::Stats]
 - Dependencies changed since 2.001001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +5 ↑1 -2 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1
 - test: +1 (recommends: ↑2)

 [Minor]
 - Stop requiring 5.008 and utf8 when we don't really need it.

2.001001 2014-08-29T10:07:27Z
 [Bugfix]
 - avoid no-build-dir fails.

 [Dependencies::Stats]
 - Dependencies changed since 2.001000, see misc/*.deps* for details
 - test: ↑1

2.001000 2014-08-22T18:50:50Z
 [00 Minor]
 - Rewrite dump_config in terms of new Utility.

 [Dependencies::Stats]
 - Dependencies changed since 2.000010, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +1

2.000010 2014-08-19T01:26:37Z
 [Code]
 - Remove unused Readonly stuff.

 [Dependencies::Stats]
 - Dependencies changed since 2.000003, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - runtime: -1
 - test: +1 ↑1

 [Documentation]
 - Add very experimental QUICK REFERENCE sections.

 [Tests]
 - reimplemented in terms of my newish tester util
 - some tests split into smaller tests.

2.000003 2014-08-16T00:11:28Z
 [00 Trivial]
 - CPANDAY!
 - metdata changes.
 - tests augmented.
 - whitespace adjusted.
 - some POD tweaked.

 [Dependencies::Stats]
 - Dependencies changed since 2.000002, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: +1 -3 (recommends: ↑1 -1)

 [Documentation]
 - Heading order and notation slightly improved.

 [Metadata]
 - Migrate to kentnl org

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - Update reportprereqs

2.000002 2014-07-13T12:27:14Z
 [00 Trivial]
 - Documentation/Spelling errors ( closes #2 )

 [Dependencies::Stats]
 - Dependencies changed since 2.000001, see misc/*.deps* for details
 - develop: (suggests: ↑2)

2.000001 2014-07-12T19:31:53Z
 [00 Trivial]
 - Misc packaging cleanups.

 [Dependencies::Stats]
 - Dependencies changed since 2.000000, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - test: (recommends: ↑2)

2.000000 2014-02-02T05:17:53Z
 [00 Major]
 - Moving to EUMM ( May break non-cpan toolchains that work by HAND )
 - Mantissa length shortening to 6 for more boring versions. ( Primary reason for 1 → 2 )
 - Dependencies on EUMM/Test::More downgraded to recommends where possible.
 - 5.008 minimum

 [Dependencies::Aggregate]
 - 5.8 : Before = 205, After = 204
 - 5.10 : Before = 183, After = 183
 - 5.19 : Before = 169, After = 167
 - Will be less when Dzil drops Path::Class (somehow)

 [Dependencies::Stats]
 - Dependencies changed since 1.15000200, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +4 ↑1 -5
 - runtime: +1 ↑1
 - test: +4 ↓1 (recommends: +4)

1.15000200 2013-12-29T08:37:43Z
 [Bugfix]
 - Avoid bug with Dzil being gc'd

 [Depedencies::Noteworthy]
 - Module::Build 0.4202 → 0.4203

 [Dependencies::Stats]
 - Dependencies changed since 1.15000100, see misc/*.deps* for details

t/lib/TZil.pm
use strict;
use warnings;
package # NO PAUSE PLX
  TZil;

use Try::Tiny;
use Dist::Zilla::Tester qw( Builder );
use Params::Util qw(_HASH0);

use Sub::Exporter -setup => {
  exports => [
    test_config =>
    simple_ini  => \'_simple_ini',
  ],
  groups => [ default => [ qw( -all ) ] ]
};

sub test_config {
  my ( $conf ) = shift;
  my $args = [];
  if ( $conf->{dist_root} ) {
    $args->[0] = { dist_root => $conf->{dist_root} };
  }
  if ( $conf->{ini} ){
    $args->[1] ||= {};
    $args->[1]->{add_files} ||= {};
    $args->[1]->{add_files}->{'source/dist.ini'} = _simple_ini()->( @{ $conf->{ini} } );
  }
  my $build_error = undef;
  my $instance;
  try {
    $instance = Builder->from_config( @$args );

    if ( $conf->{build} ){
      $instance->build();
    }
  } catch {
    $build_error = $_;
  };

  if ( $conf->{post_build_callback} ) {
    $conf->{post_build_callback}->({
      error => $build_error,
      instance => $instance,
    });
  }

  if ( $conf->{find_plugin} ){
    my $plugin = $instance->plugin_named( $conf->{find_plugin} );
    if ( $conf->{callback} ){
      my $error = undef;
      my $method = $conf->{callback}->{method};
      my $args   = $conf->{callback}->{args};
      my $call   = $conf->{callback}->{code};
      my $response;
      try {
        $response = $instance->$method( @$args );
      } catch {
        $error = $_;
      };
      return $call->({
        error => $error,
        response => $response,
        instance => $instance,
      });
    } else {
      return $plugin;
    }
  }
}

sub _build_ini_builder {
  my ($starting_core) = @_;
  $starting_core ||= {};

  sub {
    my (@arg) = @_;
    my $new_core = _HASH0($arg[0]) ? shift(@arg) : {};

    my $core_config = { %$starting_core, %$new_core };

    my $config = '';

    for my $key (keys %$core_config) {
      my @values = ref $core_config->{ $key }
                 ? @{ $core_config->{ $key } }
                 : $core_config->{ $key };

      $config .= "$key = $_\n" for grep {defined} @values;
    }

    $config .= "\n" if length $config;

    for my $line (@arg) {
      my @plugin = ref $line ? @$line : ($line, {});
      my $moniker = shift @plugin;
      my $name    = _HASH0($plugin[0]) ? undef : shift @plugin;
      my $payload = shift(@plugin) || {};

      die "TOO MANY ARGS TO PLUGIN GAHLGHALAGH" if @plugin;

      $config .= '[' . $moniker;
      $config .= ' / ' . $name if defined $name;
      $config .= "]\n";

      for my $key (keys %$payload) {
        my @values = ref $payload->{ $key }
                   ? @{ $payload->{ $key } }
                   : $payload->{ $key };

        $config .= "$key = $_\n" for @values;
      }

      $config .= "\n";
    }

    return $config;
  }
}

sub _dist_ini {
  _build_ini_builder;
}

sub _simple_ini {
  _build_ini_builder({
    name     => 'DZT-Sample',
    abstract => 'Sample DZ Dist',
    version  => '0.001',
    author   => 'E. Xavier Ample <example@example.org>',
    license  => 'Perl_5',
    copyright_holder => 'E. Xavier Ample',
  });
}

1;

t/lib/TestClean.pm
package TestClean;
our $VERSION = '1.10001919';


# $Id:$
my $CLASS = __PACKAGE__;
use base 'Test::Builder::Module';
@EXPORT = qw( is_clean );

sub pp {
    require Data::Dumper;
    local $Data::Dumper::Terse  = 1;
    local $Data::Dumper::Useqq  = 1;
    local $Data::Dumper::Indent = 0;
    return Data::Dumper::Dumper(shift);
}

sub is_clean($;$) {
    my $file = shift;
    my $msg  = shift;
    $msg ||= "Cleanlyness for $file";
    my $tb   = $CLASS->builder;
    my $fh;

    if ( not open $fh, '<', $file ) {
        my $o = $tb->ok( 0, $msg );
        $tb->diag("Loading $file Failed");
        return $o;
    }
    while ( my $line = <$fh> ) {

        # Tailing Whitespace is pesky
        if ( $line =~ qr/\h$/m ) {
            my $o = $tb->ok( 0, $msg );
            $tb->diag( "\n\n\\h found on end of line $. in $file\n" . pp($line) . "\n" );
            return $o;
        }

        # Tabs are teh satan.
        if ( $line =~ qr/\t/m ) {
            my $o = $tb->ok( 0, $msg );
            $tb->diag( "\\t found in line $. in $file\n" . pp($line) . "\n" );
            return $o;
        }

        # Perltidyness in teh comments
        if ( $line =~ qr/[)][{]/ ) {
            my $o = $tb->ok( 0, $msg );
            $tb->diag( ')' . "{ found in line $. in $file\n" . pp($line) . "\n" );
            return $o;
        }
    }
    close $fh;
    return $tb->ok( 1, $msg );
}

1;


t/eg/DZ2/lib/DZ2.pm
use strict;
use warnings;

package DZ2;

# ABSTRACT: this is a sample package for testing Dist::Zilla;

sub main {
  return 1;
}

1;

__END__

=head1 NAME

DZ2

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Generating and Populating 'provides' in your META.yml",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-MetaProvides",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Plugin::MetaProvides",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dist::Zilla::Role::MetaProvider" => 0,
    "Hash::Merge::Simple" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "MooseX::Types" => 0,
    "MooseX::Types::Moose" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::Role::Plugin" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Path::Tiny" => 0,
    "Scalar::Util" => 0,
    "Test::DZil" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "lib" => 0
  },
  "VERSION" => "2.002005",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/01-Provider/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Dist::Zilla::Role::MetaProvider" => 0,
  "Dist::Zilla::Role::Plugin" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Hash::Merge::Simple" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "MooseX::Types" => 0,
  "MooseX::Types::Moose" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Test::DZil" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "lib" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    Zilla accessors MetaProvides MetaProvider Plugin ModVersion undef ProviderObject yml ini metadata mungers hashref plugins parseable YAML metaproviders Plugins ProvideRecord's xdg DAGOLDEN MetaNoIndex plugin SUBTYPES recognised
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

my @exempt_subs = (
  qw(
    MooseX::Types::subtype MooseX::Types::where MooseX::Types::as MooseX::Types::ModVersion MooseX::Types::ProviderObject
    )
);
for my $exempt_sub (@exempt_subs) {
  $bundle->add_or_append_policy_field( 'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => $exempt_sub ), );
}

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');

#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
$bundle->remove_policy('Documentation::RequirePodLinksIncludeText');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.008.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4203"
  },
  "configure_requires" => {
    "Module::Build" => "0.4203"
  },
  "dist_abstract" => "Generating and Populating 'provides' in your META.yml",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-MetaProvides",
  "dist_version" => "1.15000200",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::MetaProvides",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dist::Zilla::Role::MetaProvider" => 0,
    "Hash::Merge::Simple" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "MooseX::Types" => 0,
    "MooseX::Types::Moose" => 0,
    "Readonly" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dist::Zilla::Util::Test::KENTNL" => "0.01000004",
    "Scalar::Util" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001002",
    "lib" => 0
  }
);


my %fallback_build_requires = (
  "Dist::Zilla::Util::Test::KENTNL" => "0.01000004",
  "Module::Build" => "0.4203",
  "Scalar::Util" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "1.001002",
  "lib" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
